modules/file-system/filestore/main.tf

```diff
@@ -79,4 +78,26 @@ resource "google_filestore_instance" "filestore_instance" {
       delete = "1h"
     }
   }
+
+  lifecycle {
+    precondition {
+      condition = (
+        var.reserved_ip_range == null ||
+        var.connect_mode == "DIRECT_PEERING" && can(cidrhost(var.reserved_ip_range, 0)) && contains(["24", "29"], try(split("/", var.reserved_ip_range)[1], "")) ||
+        var.connect_mode == "PRIVATE_SERVICE_ACCESS" && var.private_vpc_connection_peering != null && var.reserved_ip_range != null
+      )
```

Comment:
Do we really want to abandon all checks if `var.reserved_ip_range == null`?, it contradicts the check on the last line `var.reserved_ip_range != null`.

I think this will be much more understandable and maintainable if we split it into multiple validation blocks each with their own error message ([example of multiple conditions](https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/c39cbb016627014b3f8bb073d63124bb37bac67f/community/modules/compute/gke-node-pool/main.tf#L149-L162))
```
      condition = (
        var.connect_mode == "DIRECT_PEERING" ? can(cidrhost(var.reserved_ip_range, 0)) : true
      )
      # maybe even consider checking for the var.filestore_tier, 
      # but if so it should be done such that it fails open,
      # aka only make statements about know tiers
      condition = (
        var.connect_mode == "DIRECT_PEERING" ? contains(["24", "29"], try(split("/", var.reserved_ip_range)[1], "")) : true
      )
      condition = (
        var.connect_mode == "PRIVATE_SERVICE_ACCESS" ? var.private_vpc_connection_peering != null : true
      )
      condition = (
        var.connect_mode == "PRIVATE_SERVICE_ACCESS" ? var.reserved_ip_range != null : true
      )
```

Depending on the answer to my first question you may have to incorporate more checks for var.reserved_ip_range.

Reply:
I believe `null` is the "encouraged" value. The final `var.reserved_ip_range != null` is useless from purely logic perspective and can be removed.

https://cloud.google.com/filestore/docs/creating-instances#configure_a_reserved_ip_address_range

I personally find the code a lot _less_ readable because there's so many "else true" dead-ends.

I will see what I can do.

Reply:
Ultimately I believe the logic is easier to write and maintain (at this scope) with several ORs in a single block. I made the following changes:

- Remove `private_vpc_connection_peering`. The VPC Services Networking peering connection is largely implicit. When this is a static value (supplied as a string in blueprint), the dependency does nothing. When supplied via a module output, it only blocks in so far as the previous value blocks. It is not used in API.
- Likewise, `connect_mode` and `reserved_ip_range` are doing the blocking now in the private-service-access module. This enforces graph sufficiently

The behavior I observe is that the private-service-access module breaks whenever a service networking connection already exists for a network (e.g. default). Specifically, it breaks trying to add a new named IP range to an existing connection with named IP ranges not in the list. I believe this might reasonably be called "WAI".

In this case, the user should remove the private-service-access module from a blueprint and supply EITHER:

```yaml
  connect_mode: PRIVATE_SERVICE_ACCESS
```

OR

```yaml
  connect_mode: PRIVATE_SERVICE_ACCESS
  reserved_ip_range: name-of-range-someone-told-them-is-in-service-networking-peering
```

There is no graph ordering in these 2 cases because the scenario is explicitly considering pre-existing peerings. My recommendation is to do a follow-up PR that updates 1 or more examples with the private-service-access module and guidance to delete it (and modify connect_mode explicitly) when one already exists.

Reply:
I gave prior consideration to greater enforcement of tier level and IP range suffix. I would rather scope that out for now in favor of removing a variable that isn't doing us much good here.

Reply:
To the original comment. As someone who didn't write the code, I can say that there was significant overhead trying to understand the conditions and how they connected to this behemoth of an error message:
```
        If connect_mode is set to DIRECT_PEERING and reserved_ip_range is
        specified then it must be a CIDR IP range with suffix range size 29 for
        BASIC_HDD or BASIC_SSD tiers. Otherwise the range size must be 24. If
        connect_mode is set to PRIVATE_SERVICE_ACCESS, then the name of the
        private VPC connection peering and the name of reserved IP range must
        also be supplied.
```

Part of the value of splitting would be to provide targeted error messages for each condition.

I think with the removal of `private_vpc_connection_peering` you have greatly simplified the problem, so I will stand down from my request to split. The new logic just reads like a single condition with a few return early clauses.

You can close this comment once reading.